src/Broadway/Upcasting/UpcastingInMemoryEventStore.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Broadway\Upcasting;
+
+use Broadway\Domain\DomainEventStream;
+use Broadway\Domain\DomainMessage;
+use Broadway\EventStore\EventStore;
+use Broadway\EventStore\EventVisitor;
+use Broadway\EventStore\InMemoryEventStore;
+use Broadway\EventStore\Management\Criteria;
+use Broadway\EventStore\Management\EventStoreManagement;
+
+class UpcastingInMemoryEventStore implements EventStore, EventStoreManagement
+{
+    /**
+     * @var InMemoryEventStore
+     */
+    private $eventStore;
+
+    /**
+     * @var UpcasterChain
+     */
+    private $upcasterChain;
+
+    public function __construct(InMemoryEventStore $eventStore, UpcasterChain $upcasterChain)
```

Comment:
I like how you changed it to be a wrapper instead of adjusting existing EventStoers, but ideally we'd just have an `UpcastingMemoryStore` that wraps an `EventStore`, is there a specific reason you made this specific for the InMemory implementation here?

Reply:
For now I have preferred decorate original `InMemoryEventStore` not to make my feature "invasive". However I'm agre with you, in this way I should decorate also `EventStore` (DBAL). Another convenient way could be to move the upcasting behaviour in the `EventSourcingRepository`. By decoration or adding the `UpcasterChain` in the constructor. After all `UpcasterChain` is only a registry, and if it is empty, will do not nothing

Reply:
I think you misunderstood, I don't think there's any need for a storage specific UpcasterEventStore.

```php

final class UpcastingEventStore implements EventStore, EventStoreManagement
{
  public function __construct(EventStore $eventStore, UpcasterChain $upcasterChain) {}

  // [..] the rest of the methods just like you have them.
}
```